### PR TITLE
lapack/gonum: unify parameter checks

### PR DESCRIPTION
**Pull request by vladimir-ch**
Updates #680 

Please take a look.

### CODE REVIEW

**Comment by kortschak** [comment]
Something went wrong here.

**Comment by vladimir-ch** [comment]
Yes but not sure what. Let's see if the force push fixes it.

File: lapack/gonum/dlasq3.go [hunk 1]
```diff
@@ -13,6 +13,17 @@ import "math"
 // Dlasq3 is an internal routine. It is exported for testing purposes.
 func (impl Implementation) Dlasq3(i0, n0 int, z []float64, pp int, dmin, sigma, desig, qmax float64, nFail, iter, nDiv int, ttype int, dmin1, dmin2, dn, dn1, dn2, g, tau float64) (
 	i0Out, n0Out, ppOut int, dminOut, sigmaOut, desigOut, qmaxOut float64, nFailOut, iterOut, nDivOut, ttypeOut int, dmin1Out, dmin2Out, dnOut, dn1Out, dn2Out, gOut, tauOut float64) {
+	switch {
+	case i0 < 0:
+		panic("lapack: i0 < 0")
+	case n0 < 0:
+		panic("lapack: n0 < 0")
```
Comment (btracey) [inline lapack/gonum/dlasq3.go]: nLT0?

File: lapack/gonum/dlasq3.go [hunk 2]
```diff
@@ -13,6 +13,17 @@ import "math"
...
+	case len(z) < 4*n0:
+		panic(shortZ)
+	case pp != 0 && pp != 1 && pp != 2:
+		panic("lapack: bad value of pp")
```
Comment (btracey) [inline lapack/gonum/dlasq3.go]: make this a string also?

File: lapack/gonum/dlasq4.go [hunk 3]
```diff
@@ -12,6 +12,17 @@ import "math"
 //
 // Dlasq4 is an internal routine. It is exported for testing purposes.
 func (impl Implementation) Dlasq4(i0, n0 int, z []float64, pp int, n0in int, dmin, dmin1, dmin2, dn, dn1, dn2, tau float64, ttype int, g float64) (tauOut float64, ttypeOut int, gOut float64) {
+	switch {
+	case i0 < 0:
+		panic("lapack: i0 < 0")
```
Comment (btracey) [inline lapack/gonum/dlasq4.go]: This one as well?

File: lapack/gonum/dlarf.go [hunk 4]
```diff
@@ -23,18 +23,36 @@ import (
 // Dlarf is an internal routine. It is exported for testing purposes.
 func (impl Implementation) Dlarf(side blas.Side, m, n int, v []float64, incv int, tau float64, c []float64, ldc int, work []float64) {
 	applyleft := side == blas.Left
```
Comment (kortschak) [inline lapack/gonum/dlarf.go]: Move this lower?

File: lapack/gonum/dlascl.go [hunk 5]
```diff
@@ -17,16 +17,36 @@ import (
 //
 // Dlascl is an internal routine. It is exported for testing purposes.
 func (impl Implementation) Dlascl(kind lapack.MatrixType, kl, ku int, cfrom, cto float64, m, n int, a []float64, lda int) {
-	checkMatrix(m, n, a, lda)
-	if cfrom == 0 {
-		panic(zeroDiv)
+	switch kind {
+	default:
```
Comment (kortschak) [inline lapack/gonum/dlascl.go]:
```
default:
    panic("invalid kind")
case 'H', 'B', 'Q', 'Z': // See dlascl.f.
    panic("not implemented")
```

Comment (vladimir-ch) [reply lapack/gonum/dlascl.go]: Good idea, thanks.

File: lapack/gonum/dlarf.go (hunk 4 above)
Comment (vladimir-ch) [reply lapack/gonum/dlarf.go]: Thanks, this was an oversight.

**Pull request merged**